### DISCUSSION TITLE
Suggest a custom component 🎡

### DISCUSSION THREAD

**Discussion by melohagan** [Budibase/budibase#7858] [Ideas]
With the upcoming release of custom components, make your suggestions here!

Kanban board? Interactive calendar? Anything at all you would find useful, fire it in!

The list of current custom components can be found here: https://github.com/Budibase/plugins

**Comment by poirazis** [comment]
I would like a Tab Control !

**Reply by JayP718** [reply]
+1

**Comment by pbarmak** [comment]
I would love to see inline table editing (as an alternative to separate form screens) and the concept of sub-forms that link to their parent form via an ID field.

**Reply by joebudi** [reply]
hey @pbarmak . Out of curiosity, would an example of a sub-form be a task, with subtasks?

**Reply by pbarmak** [reply]
@joebudi, yes, that's a great example. Several sub-tasks for a single parent task. Great for ticketing systems as well (multiple tasks for a ticket to be resolved). We also have an invoicing system where a parent invoice has multiple invoice line items. It would be great to have those line items as a sub-form and, with the inline table editing feature I'd love to have, that sub-form can be updated right in the parent form screen (no need to jump around screens). Having it as a sub-form would also be useful for developers, as we can just call that same sub-form up into multiple screens of parent forms (admin vs non-admin screens, things like that), no need to copy/paste components across screens and having to remember to update each one every time we have a new field/filter we want to add. I love Budibase, these additions would be amazing. Same with kanban boards and other features, I can see a ton of benefits with these new features.

**Reply by joebudi** [reply]
Thanks for the write-up and your kind words. I'll feed this back to the product team.

**Reply by tmdoit** [reply]
Hi @joebudi,
is this in you roadmap?

**Comment by kmeesters** [comment]
Hi! A relative new user to Budibase here, but very happy thus far!
For the components: here are some ideas that I ran into in a previous project:

- A 'tree' component: something that allows you to work with nested data (for example where DB has 'parent' field. Use-case: think about navigation for example, or nested content. 
- A calendar view for data, (and possibly to also extend it to a Gantt chart). Ideally where you can drag and drop (and thus update the DB) if allowed. This allows for example to build booking systems or scheduling systems.
- KanBan: yes! Ideal for project management or task tracking applications. While you could create it manually with repeaters and card, it is cumbersome.

And I guess there are many existing javascript/react UI components that can serve as inspiration. But the above ones are things I missed in previous project indeed.

**Comment by melohagan** [comment]
Number counter (with plus/minus) form field: https://github.com/Budibase/budibase/discussions/7990

**Reply by melohagan** [reply]
I've added a plugin for a number spinner form field here: https://github.com/melohagan/budibase-component-number-spinner

Not quite a plus/minus field but might be nice for some use-cases.

**Comment by feurer98** [comment]
A picture gallery:

In combination with the comment plugin, a rating of images could then be made. Or a selection of images could be made. I am now thinking about for a workflow of photographers.

Calendar Plugin

More comfortable input of start and end over several days to plan trips or rent objects.

**Comment by rafikApp** [comment]
A signature component

**Reply by melohagan** [reply]
Community contribution! https://github.com/FlaminWrap/budibase-signature-component

Custom plugins docs: https://docs.budibase.com/docs/custom-plugin

**Comment by dosandco** [comment]
What a cool post!  The ideas we would love to see are:

- A newsfeed / timeline component (perhaps along the line of this free one https://getstream.io/activity-feeds/ui-kit/#download-kits) for us to be able to show updates to our teams

- A gantt chart component (where you can show project tasks and dependencies) and perhaps click through from bars to a detail/side window.

- Some sort of 'swipe pile' container with an action on swipe left or swipe right.  This is a suuuuuuper efficient way of getting things approved/rejected on mobile and, if it's a container, we could build some really useful use cases.  Could also assist with the photography request above https://github.com/Budibase/budibase/discussions/7858#discussioncomment-3814395.

- A stacked area / stacked bar chart - just like the awesome ones we already have, but with the ability to add more than one category on the same chart and show them stacked.

- Nested tabs - at present, it isn't possible to nest tab sets inside other tab sets - it would be great to be able to do so (and great to be able to activate a specific tab on a page via URL parameters).

- A hierarchy chart container - for showing organisation charts / corporate structures - being able to set this up so that it is possible to have containers as 'nodes' (so they can be used as custom content) and be able to label the lines between them and have actions on clicking those lines (to find out more about the relationships between, say, entities or management lines) - the one here works really nicely! https://theorg.com/org/google 

- A matrix / pivot table view, whereby two datasets can be input, allowing you dynamically to display relationships/overlaps in each box.  One use case would be departments as rows, and years as columns (showing sales/staff or similar), and the ability to then show information dynamically (at present, column values must be set manually).  This would also enable dynamic balance sheets/profit and loss reports.  The ability to make the first x rows / first x columns 'sticky' where a table is bigger than the view would really help with this, too.  The equivalent for this would be a matrix input in the form section - again, allowing for very efficient data entry (though this probably duplicates the editable table fields above).

- A sidebar filter (like an eCommerce one) whereby you can filter a set of cards based on selected columns and bind the two components together.  Use case here might be for large corporate directories, being able to filter intuitively by department/seniority etc (setting check-boxes or similar) would be much more user friendly than the current Dynamic Filter.

**Reply by poirazis** [reply]
Hwy @dosandco 
if you are willing to use custom plugins, you can achieve nested Tab Groups as deep as you want with the SuperTabs component.
https://github.com/poirazis/bb-component-SuperTabs

https://user-images.githubusercontent.com/80512170/220354880-e8550aab-3e9b-4053-9cca-db73a848adb8.mov

<img width="245" alt="Screenshot 2023-02-21 at 3 13 50 PM" src="https://user-images.githubusercontent.com/80512170/220354892-d752e369-e77d-4fc7-95c3-aa0d4ee693be.png">

**Comment by dosandco** [comment]
Perfect!  That is just what I was looking for.  Amazing plugin, thank you.

**Comment by 5hanth** [comment]
I would like to see a Block component that exposes window.web3 object so that any child component can access that object to perform web3 interactions.
